starfish/recipe/recipe.py

```diff
@@ -0,0 +1,159 @@
+from typing import (
+    AbstractSet,
+    Any,
+    Mapping,
+    MutableMapping,
+    MutableSequence,
+    Sequence,
+    Set,
+)
+
+from .filesystem import FileProvider, FileTypes
+from .runnable import Runnable
+
+
+class Execution:
+    """Encompasses the state of a single execution of a recipe."""
+    def __init__(
+            self,
+            runnable_sequence: Sequence[Runnable],
+            output_runnables: Sequence[Runnable],
+            output_paths: Sequence[str],
+    ) -> None:
+        self._runnable_sequence = iter(runnable_sequence)
+        self._output_runnables = output_runnables
+        self._output_paths = output_paths
+
+        # build a set of all the runnables we need to arrive at the outputs, and a mapping from each
+        # runnable to its dependents.
+        needed_runnables: Set[Runnable] = set()
+        runnable_dependents: MutableMapping[Runnable, Set[Runnable]] = dict()
+        for output_runnable in self._output_runnables:
+            Execution.build_graph(output_runnable, needed_runnables, runnable_dependents)
+        self.needed_runnables: AbstractSet[Runnable] = needed_runnables
+        self.runnable_dependents: Mapping[Runnable, AbstractSet[Runnable]] = runnable_dependents
+
+        # completed results
+        self._completed_runnables: Set[Runnable] = set()
+        self._completed_results: MutableMapping[Runnable, Any] = dict()
+        self._incomplete_outputs: Set[Runnable] = set(self._output_runnables)
+
+    @property
+    def complete(self) -> bool:
+        return len(self._incomplete_outputs) == 0
+
+    def run_one_tick(self) -> None:
+        """Run one tick of the execution graph."""
+        while True:
+            candidate_runnable = next(self._runnable_sequence)
+
+            if candidate_runnable in self.needed_runnables:
```

Comment:
what exactly does self.needed_runnables mean in this context?

Reply:
As discussed offline, this is undesirable, and as such, removed.